### PR TITLE
Use category dtype for type in PropertyGraph

### LINKED ISSUE #2399: [FEA]  Use category dtype in PropertyGraph DataFrames
**Describe the solution you'd like and any additional context**

We should use  category dtype for `_TYPE_` column rather than string for `_edge_prop_dataframe` and `_vertex_prop_dataframe`. 

This will have following benefits: 

1. Saves on memory Consumption (See below example for motivation)
2. Faster because of numerical dtypes
3. Also aligns us to Graph Frameworks like `DGL` who also use category like dtypes to save `_TYPE_` under the hood. 

Memory Usage Motivational example: 

```python
edge_prop_dataframe



_SRC_ | _DST_ | _EDGE_ID_ | _TYPE_ | user_id_1 | user_id_2 | merchant_id | stars | relationship_type | user_id | volume | time | card_num
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
89216 | 78634 | 0 | referrals | 89216 | 78634 | 11 | 5 | <NA> | <NA> | <NA> | <NA> | <NA>
89216 | 78634 | 0 | referrals | 89216 | 78634 | 11 | 5 | <NA> | <NA> | <NA> | <NA> | <NA>
89216 | 78634 | 0 | referrals | 89216 | 78634 | 11 | 5 | <NA> | <NA> | <NA> | <NA> | <NA>

``` 

```python
get_size(edge_prop_dataframe['_TYPE_'].memory_usage())
'3.0 GB'
```

```python
get_size(edge_prop_dataframe['_TYPE_'].astype('category').memory_usage())
'1.17 GB'
```

### PR DESCRIPTION
**Pull request by eriknw**
Closes #2399

This is ~not yet~ *now* implemented for MG.

### CODE REVIEW

**Comment by VibhuJawa** [comment]
> So, should we allow users to add types to the property graph? This allows types to exist that have no data yet. I'm inclined to say "not yet

I think "not yet" is the correct assumption.  I think the main focus is to make the PGs efficient (both memory and computation) once the graph is created  and not optimize the graph creation process.

**Review by VibhuJawa — APPROVED** [review]
LGTM to me. Thanks for working on this Erik.

**Review by rlratzel — COMMENTED** [review]
LGTM, I just had one possible suggestion:

File: python/cugraph/cugraph/structure/property_graph.py [hunk 1]
```diff
@@ -416,7 +428,37 @@ def add_vertex_data(self,
         tmp_df = dataframe.copy(deep=True)
         tmp_df[self.vertex_col_name] = tmp_df[vertex_col_name]
         # FIXME: handle case of a type_name column already being in tmp_df
-        tmp_df[self.type_col_name] = type_name
+
+        # Add `type_name` to the categorical dtype if necessary
```
[the remark this answers is not in the document]
Comment (eriknw) [reply python/cugraph/cugraph/structure/property_graph.py:432]: Oh, that's reasonable.  I'll do that.  Good suggestion.

Heh, I thought you were commenting on something else, and I had this reaction:
> This "wrinkle" was intentional and is a matter of aspiration.
> 
> I aspire to write canonical pandas code.  I think `cudf` aspires to be like canonical pandas.
> 
> I also acknowledge that pandas, dask, cudf, and dask_cudf are _all_ different.  APIs are different, behaviors are different, bugs/workarounds are different, performance characteristics are different, etc.  Doing "least common denominator" development may be similarly sloppy and limiting.  So, I also aspire to refactor this code to better accommodate the differences between the libraries.

[the remark this answers is not in the document]
Comment (eriknw) [reply python/cugraph/cugraph/structure/property_graph.py:432]: Done.

**Review by rlratzel — APPROVED** [review]

**Comment by rlratzel** [comment]
rerun tests

**Comment by alexbarghi-nv** [comment]
👍

**Comment by BradReesWork** [comment]
rerun tests

**Comment by BradReesWork** [comment]
rerun tests

**Comment by eriknw** [comment]
Using categoricals for the type changes the behavior of `renumber_edges_by_type` and `renumber_vertices_by_type`.  Previously, we would sort using on the _name_ of the type.  Now, we sort on the categorical value of the type, which is not alphabetical.

This probably isn't a problem, since we give the start and end for each type, and otherwise the behavior is well-behaved and correct.

**Comment by BradReesWork** [comment]
rerun tests

**Comment by BradReesWork** [comment]
rerun tests

**Comment by BradReesWork** [comment]
rerun tests

**Pull request merged**